### PR TITLE
Edit display typing

### PR DESCRIPTION
**Pull request by MarkvdVorst**
Part of refactoring the helper.service.ts

Added types to the edit display and helper.service.ts

### CODE REVIEW

File: src/app/shared/services/helper.service.ts [hunk 1]
```diff
@@ -1,44 +1,42 @@
...
 
-  setButtonHtml(report: any, button: any, type: string, showConverted: boolean): void {
-    report.showConverted = showConverted;
-    button.target.title = `Convert to ${type}`;
-    button.target.innerHTML = type;
+  setButtonHtml(checkpoint: Checkpoint, button: MouseEvent, type: string, showConverted: boolean): void {
+    checkpoint.showConverted = showConverted;
+    const target: HTMLElement | null = button.target as HTMLElement;
+    target.title = `Convert to ${type}`;
+    target.innerHTML = type;
```
Comment (MatthijsSmets) [inline src/app/shared/services/helper.service.ts:39]: This function should not be here anymore, this button with its functionality should be its own component

Comment (MarkvdVorst) [reply src/app/shared/services/helper.service.ts]: I added the change. I also moved the convertMessage method fromthe helper service to the edit-display component as it was not used anywhere else.

File: src/app/report/edit-display/encoding-button/encoding-button.component.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  setButtonHtml(checkpoint: Checkpoint, button: MouseEvent, type: string, showConverted: boolean): void {
+    checkpoint.showConverted = showConverted;
+    const target: HTMLElement | null = button.target as HTMLElement;
+    target.title = `Convert to ${type}`;
+    target.innerHTML = type;
+  }
+
+  protected readonly ReportUtil = ReportUtil;
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/encoding-button/encoding-button.component.ts:50]: This has to be refactored, just use a class variable for the text of the button

File: src/app/report/edit-display/encoding-button/encoding-button.component.html [hunk 3]
```diff
@@ -0,0 +1,3 @@
+@if (ReportUtil.isCheckPoint(selectedNode) && selectedNode.encoding === 'Base64') {
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/encoding-button/encoding-button.component.html:1]: This if statement can be simplified to a class variable that is updated in `ngOnChanges`

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

File: src/app/report/edit-display/encoding-button/encoding-button.component.ts (hunk 2 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/encoding-button/encoding-button.component.ts]: added the change

File: src/app/report/edit-display/encoding-button/encoding-button.component.html (hunk 3 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/encoding-button/encoding-button.component.html]: added the change

File: src/app/report/edit-display/encoding-button/encoding-button.component.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  selector: 'app-encoding-button',
+  standalone: true,
+  imports: [],
+  templateUrl: './encoding-button.component.html',
+  styleUrl: './encoding-button.component.css',
+})
+export class EncodingButtonComponent implements OnChanges {
+  @Input({ required: true }) selectedNode!: ReportOrCheckpoint;
+  @Input({ required: true }) editor!: CustomEditorComponent;
+  buttonType: string = 'Base64';
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/encoding-button/encoding-button.component.ts:16]: Does it always start at Base64?

File: src/app/report/edit-display/encoding-button/encoding-button.component.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+@Component({
+  selector: 'app-encoding-button',
+  standalone: true,
+  imports: [],
+  templateUrl: './encoding-button.component.html',
+  styleUrl: './encoding-button.component.css',
+})
+export class EncodingButtonComponent implements OnChanges {
+  @Input({ required: true }) selectedNode!: ReportOrCheckpoint;
+  @Input({ required: true }) editor!: CustomEditorComponent;
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/encoding-button/encoding-button.component.ts:15]: This makes the flow of data sharing between components very unlogical IMO, right now this flow looks a little like:
```mermaid
flowchart TD
    A[edit-display.component] --> B[encoding-button.component];
    A <--> C[custom-editor.component]
    B --> C
```
however it should be:
```mermaid
flowchart TD
    A[edit-display.component] <--> B[encoding-button.component];
    A <--> C[custom-editor.component]
```
I think it makes more sense to have all data sharing go from the parent to the child, so in this case it's best to have the encoding-button.component have an output so the parent can do with that output as it chooses, this way our components have a tight cohesion but are loosely coupled.

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

File: src/app/report/edit-display/edit-display.component.ts [hunk 6]
```diff
@@ -85,22 +90,33 @@ export class EditDisplayComponent {
...
+      this.editor.setNewReport(this.convertMessage(this.selectedNode));
+    }
     this.rerunResult = undefined;
     this.displayReport = true;
   }
 
-  changeEncoding(button: any): void {
-    this.editor.setNewReport(this.helperService.changeEncoding(this.report, button));
+  convertMessage(checkpoint: Checkpoint): string {
+    let message: string = checkpoint.message === null ? '' : checkpoint.message;
```
Comment (ricklue) [inline src/app/report/edit-display/edit-display.component.ts:106]:
```suggestion
    let message: string = checkpoint.message ?? '';
```
I think this is a cleaner way to write this line of code, right?

File: src/app/report/edit-display/encoding-button/encoding-button.component.ts (hunk 4 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/encoding-button/encoding-button.component.ts:16]: Yes, if the encoding is set to base64 it will always start as base64

File: src/app/report/edit-display/encoding-button/encoding-button.component.ts (hunk 5 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/encoding-button/encoding-button.component.ts]: I moved the this.editor.setNewReport(message) call to a method in the edit-display component which gets called on te output attribute eventemitter from the encoding-button component.

File: src/app/report/edit-display/edit-display.component.ts (hunk 6 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Yes, I had not noticed that line yet. I added the change.

File: src/app/report/edit-display/edit-display.component.ts [hunk 7]
```diff
@@ -103,7 +103,7 @@ export class EditDisplayComponent {
   }
 
   convertMessage(checkpoint: Checkpoint): string {
-    let message: string = checkpoint.message === null ? '' : checkpoint.message;
+    let message: string = checkpoint.message ?? '';
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/edit-display.component.ts:106]: But if message is an empty string, this whole function does nothing right?

Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts:106]:
```suggestion
    let message: string = checkpoint.message;
```
I'll just change it to this for thread safety and leave it at that..

Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Added the change.

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

File: src/app/report/edit-display/edit-display.component.ts [hunk 8]
```diff
@@ -85,22 +90,33 @@ export class EditDisplayComponent {
...
+    let message: string = checkpoint.message;
+    if (checkpoint.encoding == 'Base64') {
+      message = btoa(message);
+    }
+    return message;
   }
 
   rerunReport(): void {
-    const reportId: number = this.report.storageId;
+    const node: ReportOrCheckpoint = this.selectedNode;
```
[the remark this answers is not in the document]
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts:114]: I added this after you asked me to avoid a race condition/shared mutable state for the code. So I added a local variable so it avoids a shared mutable state in case selectedNode gets changed while the method is running.

[the remark this answers is not in the document]
Comment (MatthijsSmets) [reply src/app/report/edit-display/edit-display.component.ts:114]: You're right, my bad.

File: src/app/report/edit-display/edit-display.component.html [hunk 9]
```diff
@@ -66,16 +66,14 @@
           (click)="openDifferenceModal('save')"
         ><i class="fa fa-save"></i></button>
       }
-      @if (report.encoding === 'Base64') {
-        <button class="btn btn-info" title="Convert to Base64" (click)="changeEncoding($event)">Base64</button>
-      }
+      <app-encoding-button [selectedNode]="selectedNode" (updatedMessageEvent)="setNewReport($event)"></app-encoding-button>
 
-      @if (report.stub) {
+      @if (selectedNode && selectedNode.stub) {
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/edit-display.component.html:71]: This should only be visible if report is a checkpoint

File: src/app/report/edit-display/edit-display.component.ts [hunk 10]
```diff
@@ -57,15 +59,17 @@ import { View } from '../../shared/interfaces/view';
...
 })
 export class EditDisplayComponent {
+  protected readonly ReportUtil = ReportUtil;
   @Input() id: string = '';
   @Input() containerHeight!: number;
   @Input({ required: true }) currentView!: View;
   @Input() newTab: boolean = true;
   @Output() saveReportEvent: Subject<any> = new Subject<any>();
-  @Output() closeReportEvent: Subject<void> = new Subject<void>();
+  @Output() closeReportEvent: Subject<Report> = new Subject<Report>();
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/edit-display.component.ts:72]: Why was this changed from void?
The selectedNode is passed by it's parent, so you don't have to pass which report is closed here as the parent already knows.

File: src/app/report/edit-display/edit-display.component.ts [hunk 11]
```diff
@@ -111,33 +126,59 @@ export class EditDisplayComponent {
...
-    this.saveChanges((event.target as HTMLOptionElement).value);
+    this.saveChanges(Number.parseInt((event.target as HTMLOptionElement).value));
   }
 
   openDifferenceModal(type: ChangesAction): void {
+    const node: Report | Checkpoint | undefined = this.selectedNode;
+    if (!node) {
+      this.toastService.showDanger('Could not find report to open difference modal');
+      return;
+    }
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/edit-display.component.ts:174]: I see these two lines used in multiple methods, maybe it's better to move this to a method that just returns a checkpoint or a report and shows a toast if it's null or undefined?

File: src/app/shared/components/display-table/metadata-table.component.ts [hunk 12]
```diff
@@ -16,7 +16,7 @@ import { ReportUtil } from '../../util/report-util';
 export class MetadataTableComponent implements OnChanges {
   protected readonly ReportUtil = ReportUtil;
 
-  @Input({ required: true }) report!: Report | Checkpoint;
+  @Input({ required: true }) report?: Report | Checkpoint;
```
Comment (MatthijsSmets) [inline src/app/shared/components/display-table/metadata-table.component.ts:19]: If it's required, then it can't be undefined

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

File: src/app/report/edit-display/edit-display.component.html [hunk 13]
```diff
@@ -66,16 +66,14 @@
           (click)="openDifferenceModal('save')"
         ><i class="fa fa-save"></i></button>
       }
-      @if (report.encoding === 'Base64') {
-        <button class="btn btn-info" title="Convert to Base64" (click)="changeEncoding($event)">Base64</button>
-      }
+      <app-encoding-button [selectedNode]="selectedNode" (updatedMessageEvent)="setNewReport($event)"></app-encoding-button>
 
-      @if (report.stub) {
+      @if ((ReportUtil.isReport(selectedNode) || ReportUtil.isCheckPoint(selectedNode)) && selectedNode.stub) {
```
[the remark this answers is not in the document]
Comment (MatthijsSmets) [reply src/app/report/edit-display/edit-display.component.html]: My bad, this should only be type checkpoint

[the remark this answers is not in the document]
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.html]: Both report and checkpoint have it so just checking for undefined is good enough. Added the change

File: src/app/report/edit-display/edit-display.component.ts (hunk 10 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts:72]:
```ts
closeReport(removeReportFromTree: boolean): void {
    this.displayReport = false;
    this.editingRootNode = false;
    this.editingChildNode = false;
    if (removeReportFromTree) {
      this.closeReportEvent.next(this.report);
    }
    this.editor.setNewReport('');
  }
  ```
  This was the code before the change. It gave the this.report which was of type any to the closeReportEvent. I don't know why angular allowed this to be a void, since a variable is being send anyway.
As it stands, the parent does not ask for the current report when removing but does have access to the current report. 

I'll make it so the closeReport method actually calls for the current report instead of asking for one from the event.

File: src/app/report/edit-display/edit-display.component.ts (hunk 11 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts:174]: That's a good idea. I'll add the change.

File: src/app/report/edit-display/edit-display.component.html (hunk 13 above)
[the remark this answers is not in the document]
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.html]: Added the change

File: src/app/report/edit-display/edit-display.component.ts [hunk 14]
```diff
@@ -85,22 +90,33 @@ export class EditDisplayComponent {
     private errorHandler: ErrorHandling,
   ) {}
 
-  showReport(report: Report): void {
+  showReport(node: ReportOrCheckpoint): void {
```
[the remark this answers is not in the document]
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Added the change

File: src/app/report/edit-display/edit-display.component.ts [hunk 15]
```diff
@@ -111,33 +127,57 @@ export class EditDisplayComponent {
...
+    if (!node) {
+      this.toastService.showDanger('Could not find report to download');
+      return;
+    }
+    let queryString: string;
+    queryString = ReportUtil.isReport(node)
+      ? node.storageId.toString()
+      : ReportUtil.isCheckPoint(node)
+        ? node.uid.split('#')[0]
+        : '';
```
[the remark this answers is not in the document]
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Added the change

File: src/app/report/edit-display/edit-display.component.html (hunk 9 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.html]: Added the change.

File: src/app/report/edit-display/edit-display.component.ts (hunk 10 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Added the change.

File: src/app/report/edit-display/edit-display.component.ts (hunk 11 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Added the change

File: src/app/shared/components/display-table/metadata-table.component.ts (hunk 12 above)
Comment (MarkvdVorst) [reply src/app/shared/components/display-table/metadata-table.component.ts:19]: Added the change

File: src/app/report/edit-display/edit-display.component.ts [hunk 16]
```diff
@@ -111,33 +127,57 @@ export class EditDisplayComponent {
...
     this.httpService.handleSuccess('Report Downloaded!');
   }
 
   selectStubStrategy(event: Event): void {
-    this.saveChanges((event.target as HTMLOptionElement).value);
+    this.saveChanges(Number.parseInt((event.target as HTMLOptionElement).value));
   }
 
   openDifferenceModal(type: ChangesAction): void {
+    const node: ReportOrCheckpoint = this.selectedNode;
```
[the remark this answers is not in the document]
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Added the change

**Comment by MarkvdVorst** [comment]
Branch contains a flacky bug that shows up sometimes for the debugTreeGuardedCopyReport command from cypress tests

File: src/app/report/edit-display/edit-display.component.ts [hunk 17]
```diff
@@ -244,8 +285,9 @@ export class EditDisplayComponent {
     }
   }
 
-  toggleToolTip(toolTip: MatTooltip): void {
-    toolTip.show();
-    setTimeout(() => toolTip.hide(), 2500);
+  setNewReport(message: string) {
+    this.editor.setNewReport(message);
```
Comment (ricklue) [inline src/app/report/edit-display/edit-display.component.ts:289]: You have made a setNewReport function, but you don't use it? Or is this only for the encoding button?

File: src/app/report/edit-display/edit-display.component.ts [hunk 18]
```diff
@@ -293,20 +285,9 @@ export class EditDisplayComponent {
...
   }
 
-  getSelectedNode(): Report | Checkpoint | undefined {
-    const node: Report | Checkpoint | undefined = this.selectedNode;
-    if (!node) {
-      this.toastService.showDanger('Selected report was undefined');
-    }
-    return node;
-  }
+  protected readonly Number: NumberConstructor = Number;
```
Comment (MatthijsSmets) [inline src/app/report/edit-display/edit-display.component.ts:292]: Please move this to the top of the file just like ReportUtil

**Review by MatthijsSmets — DISMISSED** [review]

File: src/app/shared/components/display-table/metadata-table.component.ts (hunk 12 above)
Comment (MatthijsSmets) [reply src/app/shared/components/display-table/metadata-table.component.ts:19]: Yes it should even

File: src/app/report/edit-display/edit-display.component.ts (hunk 18 above)
Comment (MarkvdVorst) [reply src/app/report/edit-display/edit-display.component.ts]: Added the change.

**Review by ricklue — APPROVED** [review]
